abi/dynamic/reflect_marshal.go

```diff
@@ -0,0 +1,162 @@
+	}
+
+	// Create a cache to avoid rebuilding types
+	typeCache := make(map[string]reflect.Type)
+
+	// Create a dynamic struct type
+	dynamicType := getReflectType(typeName, inputAbi, typeCache)
+
+	// Create an instance of the dynamic struct
+	dynamicValue := reflect.New(dynamicType).Interface()
+
+	// Unmarshal JSON data into the dynamic struct
+	if err := json.Unmarshal([]byte(jsonData), dynamicValue); err != nil {
+		return nil, fmt.Errorf("failed to unmarshal JSON data: %w", err)
+	}
+
+	// Marshal the dynamic struct using LinearCodec
+	writer := codec.NewWriter(0, consts.NetworkSizeLimit)
+	if err := codec.LinearCodec.MarshalInto(dynamicValue, writer.Packer); err != nil {
+		return nil, fmt.Errorf("failed to marshal struct: %w", err)
+	}
+
+	return writer.Bytes(), nil
+}
+
+func DynamicUnmarshal(inputAbi abi.ABI, typeName string, data []byte) (string, error) {
+	// Find the type in the ABI
+	abiType := findABIType(inputAbi, typeName)
+	if abiType == nil {
+		return "", fmt.Errorf("type %s not found in ABI", typeName)
+	}
+
+	// Create a cache to avoid rebuilding types
+	typeCache := make(map[string]reflect.Type)
+
+	// Create a dynamic struct type
+	dynamicType := getReflectType(typeName, inputAbi, typeCache)
+
+	// Create an instance of the dynamic struct
+	dynamicValue := reflect.New(dynamicType).Interface()
+
+	// Unmarshal the data into the dynamic struct
+	if err := codec.LinearCodec.Unmarshal(data, dynamicValue); err != nil {
+		return "", fmt.Errorf("failed to unmarshal data: %w", err)
+	}
+
+	// Marshal the dynamic struct back to JSON
+	jsonData, err := json.Marshal(dynamicValue)
+	if err != nil {
+		return "", fmt.Errorf("failed to marshal struct to JSON: %w", err)
+	}
+
+	return string(jsonData), nil
+}
+
+func getReflectType(abiTypeName string, inputAbi abi.ABI, typeCache map[string]reflect.Type) reflect.Type {
+	switch abiTypeName {
+	case "string":
+		return reflect.TypeOf("")
+	case "uint8":
+		return reflect.TypeOf(uint8(0))
+	case "uint16":
+		return reflect.TypeOf(uint16(0))
+	case "uint32":
+		return reflect.TypeOf(uint32(0))
+	case "uint64":
+		return reflect.TypeOf(uint64(0))
+	case "int8":
+		return reflect.TypeOf(int8(0))
+	case "int16":
+		return reflect.TypeOf(int16(0))
+	case "int32":
+		return reflect.TypeOf(int32(0))
+	case "int64":
+		return reflect.TypeOf(int64(0))
+	case "Address":
+		return reflect.TypeOf(codec.Address{})
+	default:
+		if strings.HasPrefix(abiTypeName, "[]") {
+			elemType := getReflectType(strings.TrimPrefix(abiTypeName, "[]"), inputAbi, typeCache)
+			return reflect.SliceOf(elemType)
+		} else if strings.HasPrefix(abiTypeName, "[") {
+			// Handle fixed-size arrays
+
+			sizeStr := strings.Split(abiTypeName, "]")[0]
+			sizeStr = strings.TrimPrefix(sizeStr, "[")
+
+			size, err := strconv.Atoi(sizeStr)
+			if err != nil {
+				return reflect.TypeOf((*interface{})(nil)).Elem()
+			}
+			elemType := getReflectType(strings.TrimPrefix(abiTypeName, "["+sizeStr+"]"), inputAbi, typeCache)
```

Comment:
I was debating that. The string implementation is about 3 times faster and takes less mental effort to read. If you compile the regex separately, you'll regain almost all the speed, but lose readability. I ended up compiling it separately for performance and added the actual regex as a comment where it's used. I do like my string manipulation implementation, but for an open-source product, the regex looks a bit cleaner, yes.

Reply:
Is speed necessary in a CLI?

Lose readability because it's not co-located? I happened, by chance, to comment on that in my re-review as there's no need to keep the variable declared up the top.